Import watch only pubkeys to the keypool if private keys are disabled

achow101 opened the pull request:
If the wallet has private keys disabled, allow importing public keys into the keypool. A `keypool` option has been added to `importmulti` in order to signal that the keys should be added to the keypool.

promag commented on test/functional/wallet_importmulti.py:
```diff
@@ -467,6 +467,67 @@ def run_test (self):
         import_pub2 = self.nodes[0].getaddressinfo(addr2)['pubkey']
         assert_equal(pub2, import_pub2)
 
+        # Import some public keys to the keypool of a no privkey wallet
```
Commit "Fetch keys from keypool when private keys are disabled"

Could move test to a different commit?

achow101 replied:
Done

promag commented on test/functional/wallet_importmulti.py:
```diff
@@ -467,6 +467,67 @@ def run_test (self):
 
+        # Import some public keys to the keypool of a no privkey wallet
+        self.log.info("Adding pubkey to keypool of disableprivkey wallet")
+        self.nodes[1].createwallet("noprivkeys", True)
```
Commit "Fetch keys from keypool when private keys are disabled"

Use named argument `disable_private_keys=True`?

achow101 replied:
Done

promag commented on src/wallet/rpcdump.cpp:
```diff
@@ -1198,6 +1219,7 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
             "      \"internal\": <true>                                    , (boolean, optional, default: false) Stating whether matching outputs should be treated as not incoming payments\n"
             "      \"watchonly\": <true>                                   , (boolean, optional, default: false) Stating whether matching outputs should be considered watched even when they're not spendable, only allowed if keys are empty\n"
             "      \"label\": <label>                                      , (string, optional, default: '') Label to assign to the address (aka account name, for now), only allowed with internal=false\n"
+            "      \"keypool\": <true|false>                               , (boolean, optional, default: true) If true, adds the pubkeys to the keypool if private keys are disabled for the wallet.\n"
```
Commit "Add option to importmulti add an imported pubkey to the keypool"

Default false.

achow101 replied:
Fixed

promag commented on src/wallet/rpcdump.cpp:
```diff
@@ -876,6 +877,16 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+
+        // Add to keypool only works with pubkeys
+        if (add_keypool && pubKeys.size() == 0) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Only pubkeys be imported to the keypool");
```
Commit "Add option to importmulti add an imported pubkey to the keypool"

... can be ...

achow101 replied:
Fixed

instagibbs commented on src/qt/walletmodel.cpp:
```diff
@@ -559,7 +559,7 @@ bool WalletModel::isWalletEnabled()
 bool WalletModel::privateKeysDisabled() const
 {
-    return m_wallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS);
+    return m_wallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && m_wallet->KeypoolCountExternalKeys() == 0;
```
This function should be renamed or called into from a new one.

`WalletModel::KeypoolEmptyAndPrivkeysDisabled` or something obnoxiously explicit.

achow101 replied:
Done

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -161,11 +161,15 @@ static UniValue getnewaddress(const JSONRPCRequest& request)
     }
 
-    LOCK(pwallet->cs_wallet);
+    if (!pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && !pwallet->IsLocked()) {
```
Might be time to make a `TryToTopUpKeyPool` function for when we will continue regardless. We do this in a number of places in the PR and probably elsewhere.

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -3417,7 +3423,7 @@ bool CWallet::ReserveKeyFromKeyPool(int64_t& nIndex, CKeyPool& keypool, bool fRe
             TopUpKeyPool();
 
-        bool fReturningInternal = IsHDEnabled() && CanSupportFeature(FEATURE_HD_SPLIT) && fRequestedInternal;
+        bool fReturningInternal = ((IsHDEnabled() && CanSupportFeature(FEATURE_HD_SPLIT)) || IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS))  && fRequestedInternal;
```
assumption here being using importmulti means you want to support hdsplit?

Perhaps break this into two lines:
```
bool fReturningInternal = fRequestedInternal;
fReturningInternal &= (IsHDEnabled() && CanSupportFeature(FEATURE_HD_SPLIT)) || IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS);
```

achow101 replied:
Done

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -3434,7 +3440,7 @@ bool CWallet::ReserveKeyFromKeyPool(int64_t& nIndex, CKeyPool& keypool, bool fRe
             throw std::runtime_error(std::string(__func__) + ": read failed");
         }
-        if (!HaveKey(keypool.vchPubKey.GetID())) {
+        if (!IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && !HaveKey(keypool.vchPubKey.GetID())) {
```
Belt and suspenders, we should be checking that this exists in our wallet regardless yes? Just not asking if we have the private key itself.

achow101 replied:
I changed this line to check for the pubkey in the wallet instead of the privkey.

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -3478,23 +3484,21 @@ void CWallet::ReturnKey(int64_t nIndex, bool fInternal, const CPubKey& pubkey)
 {
-    if (IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
+    LOCK(cs_wallet);
+    if (IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && KeypoolCountExternalKeys() == 0) {
```
This should be checking the keypool corresponding to `internal`, yes?

achow101 replied:
Fixed

achow101 commented:
Rebased. The first commit is still from #14019

instagibbs commented on src/interfaces/wallet.h:
```diff
@@ -272,6 +272,9 @@ class Wallet
     using WatchOnlyChangedFn = std::function<void(bool have_watch_only)>;
     virtual std::unique_ptr<Handler> handleWatchOnlyChanged(WatchOnlyChangedFn fn) = 0;
+
+    // Get the number of external keypool keys
```
generalize comment

achow101 replied:
Done

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -3311,6 +3311,22 @@ size_t CWallet::KeypoolCountExternalKeys()
+    {
+        LOCK(cs_wallet);
+        if (internal) {
+            return setInternalKeyPool.size() == 0;
```
mu-nit: add commit for KeypoolCountInternalKeys

achow101 replied:
I think that's unnecssary.

instagibbs commented:
`CWallet::CreateTransaction` needs to check for keys in change keypool even if `WALLET_FLAG_DISABLE_PRIVATE_KEYS` is set.

practicalswift commented on test/functional/wallet_importmulti.py:
```diff
@@ -446,6 +446,88 @@ def run_test (self):
+        ms = self.nodes[1].createmultisig(2, [pub1, pub2])
+        result = self.nodes[0].importmulti(
+            [{
+                'scriptPubKey' : { 'address' : ms['address']},
```
FWIW, PEP-8 whitespace nits:

```
./test/functional/wallet_importmulti.py:458:31: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:458:35: E201 whitespace after '{'
./test/functional/wallet_importmulti.py:458:45: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:459:31: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:460:26: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:482:31: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:482:35: E201 whitespace after '{'
./test/functional/wallet_importmulti.py:482:45: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:483:31: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:484:26: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:485:26: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:504:31: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:504:35: E201 whitespace after '{'
./test/functional/wallet_importmulti.py:504:45: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:505:31: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:506:26: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:507:26: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:508:27: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:523:27: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:523:31: E201 whitespace after '{'
./test/functional/wallet_importmulti.py:523:41: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:524:27: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:525:22: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:526:22: E203 whitespace before ':'
```

achow101 replied:
Fixed

Sjors commented:
Concept ACK. Lightly tested via your combined branch 3fa968ea343: I was able to import a bunch of receive and change keys. I was also able to receive coins on a `bech32` address and send from it.

For some reason when I composed the transaction it picked change address with index 1 instead of with index 0. Is that an existing rule?

What I also found confusing, though I don't know if that's this PR, or the other two or just existing weirdness, is how `dumpwallet` shows `p2sh-p2wpkh` addresses under `# addr`, even though I launched bitcoind with `-addresstype=bech32`.

achow101 commented:
> For some reason when I composed the transaction it picked change address with index 1 instead of with index 0. Is that an existing rule?

Maybe your import command was weird? It should be adding them in the order of the import command so you should be getting them in that order too.

practicalswift commented on src/wallet/rpcdump.cpp:
```diff
@@ -917,6 +928,58 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Pubkey must be a hex string");
+                }
+
+                std::vector<unsigned char> vData(ParseHex(pubkey));
```
This `vData` shadows another local variable `vData` :-)

Sjors commented:
What's a good way to inspect the wallet / keypool?

achow101 commented:
@Sjors You can use dumpwallet

Sjors commented:
@achow101 `dumpwallet` doesn't contain the bip32 derivation paths though.

achow101 commented:
I will rework this to not require #14019 as soon as I have time (in a few days)

Sjors commented:
Probably best rebased on #14491 or its predecessor.

achow101 committed:
Add a method to add a pubkey to the keypool

Introduces AddKeypoolPubkey in order to add a pubkey to the keypool

achow101 committed:
Fetch keys from keypool when private keys are disabled

When private keys are disabled, still fetch keys from the keypool
if the keypool has keys. Those keys come from importing them and
adding them to the keypool.

achow101 committed:
Add option to importmulti add an imported pubkey to the keypool

Adds a new option to importmulti where the pubkeys specified in the import
object can be added to the keypool. This only works if the wallet has
private keys disabled.

achow101 commented:
Rebased onto #14491

achow101 committed:
Test pubkey import to keypool

achow101 commented:
I've added two commits contributed by @instagibbs. The first one makes it so that things are imported in the order they are listed in the import command or in the order expanded for ranged descriptors. The second commit has watch keys imported to the keypool be marked as used when transactions are seen using them.

practicalswift commented on src/wallet/rpcdump.cpp:
```diff
@@ -804,233 +805,368 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+            // RecurseImportData() removes any relevant redeemscript/witnessscript from import_data, so we can use that to discover if a superfluous one was provided.
+            if (import_data.redeemscript) warnings.push_back("Ignoring redeemscript " + strRedeemScript + " as this is not a P2SH script.");
+            if (import_data.witnessscript) warnings.push_back("Ignoring witnessscript " + witness_script_hex + " as this is not a P2WSH script.");
+            for (const auto& privkey : import_data.privkeys) {
```
Could use `std::any_of`? :-)

MeshCollider commented:
Light utACK https://github.com/bitcoin/bitcoin/pull/14075/commits/8e6b7bd5e9aeb8efe85068de8fc38cf0b22f1185

Will review more in-depth after the dependent PRs are merged and this is rebased

Sjors commented:
I think we should make `keypoolrefill` and `TopUpKeyPool` work with just public keys.

When `WALLET_FLAG_DISABLE_PRIVATE_KEYS` is set we can skip the `EnsureWalletIsUnlocked(pwallet)` check.

`TopUpKeyPool` currently only creates random fresh keys (for legacy wallets) or derives new keys from the wallet HD seed. We could change it to first walk through `solvable` keys in the wallet.

For safety we could prevent it from putting keys without private keys into the keypool for wallets without `WALLET_FLAG_DISABLE_PRIVATE_KEYS`.

In that case we don't need the `keypool` option in `importmulti`.

This might make it easier to refactor towards a descriptor based wallet without a keypool, because we avoid adding another code path where stuff can be added to the keypool. In such a refactor a RPC method like `getnewaddress` will have to reference a descriptor. That would be the wallet receive descriptor by default. `importmulti` (or something like it) would be then be able to set the wallet receive descriptor (in case the wallet is empty).

achow101 committed:
Import public keys in order

Do public key imports in the order that they are specified in the import
or in the descriptor range.

achow101 commented:
Rebased.

I dropped the last commit as #14821 should handle that.

achow101 commented:
I've added a commit which fixed an issue where the `Request payment` button was not being enabled once keys were imported to the keypool. That commit sends a signal to the GUI every time the keypool changes so that the button will be enabled and disabled as there keys are added and removed from the keypool.

achow101 commented:
I split out 28c89d1 into it's own PR: #15225

Sjors commented:
In light of #15226 I'd like to revisit my [earlier remark](
https://github.com/bitcoin/bitcoin/pull/14075#issuecomment-446547958):

I'd like to avoid messing with the keypool inside of `importmulti`. Instead I would prefer if `TopUpKeyPool` can deal with imported keys.

> `TopUpKeyPool` currently only creates random fresh keys (for legacy wallets) or derives new keys from the wallet HD seed. We could change it to first walk through solvable keys in the wallet.

> In that case we don't need the `keypool` option in `importmulti`.

jonasschnelli commented:
Concept ACK

Ideally – but orthogonal to this PR – would be if someone could provide a xpub or range based descriptor during sethdseed (or new command) that would make the keypool act similar to the default HD keypool but derive watch only pub keys.

Sjors commented:
Tested a4bd25d, import works and GUI new address button updates as expected.

Sjors commented:
I had some difficulty combining this with #15226 (blank wallet) and made a commit that does that, poorly: https://github.com/Sjors/bitcoin/commit/8988a96e2eff8bd6344f9e93032e86742e7dc3f8
However, the blank wallet PR has since changed again, so my commit is merely useful for inspiration and some extra tests. Don't let it hold back merging here.

MeshCollider commented:
I don't think we can get both this and #14021 in to 0.18 so I'll remove this from the milestone

achow101 commented:
I've actually decided to base this on #15226. In the version of #15226, I introduced `CanGetAddresses` and `CanGenerateKeys` which I am using here for determining whether there are things in the keypool to fetch keys instead of `NoPrivkeysAndKeypoolEmpty` (which did a similar thing but was less robust). Because of this, I have also replaced `WalletModel::privateKeysDisabled` with `canGetaddresses`.

Sjors commented on src/qt/overviewpage.cpp:
```diff
@@ -184,7 +184,7 @@ void OverviewPage::setBalance(const interfaces::WalletBalances& balances)
     ui->labelImmature->setVisible(showImmature || showWatchOnlyImmature);
     ui->labelImmatureText->setVisible(showImmature || showWatchOnlyImmature);
-    ui->labelWatchImmature->setVisible(!walletModel->privateKeysDisabled() && showWatchOnlyImmature); // show watch-only immature balance
+    ui->labelWatchImmature->setVisible(walletModel->canGetAddresses() && showWatchOnlyImmature); // show watch-only immature balance
```
Same as above, but even more rare.

achow101 replied:
Reverted

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -2756,7 +2756,7 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
                 // Reserve a new key pair from key pool
-                if (IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
+                if (!CanGetAddresses(true)) {
                     strFailReason = _("Can't generate a change-address key. Private keys are disabled for this wallet.");
```
Nit: drop / tweak "Private keys are disabled for this wallet."

achow101 replied:
Changed.

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -3417,7 +3417,8 @@ bool CWallet::ReserveKeyFromKeyPool(int64_t& nIndex, CKeyPool& keypool, bool fRe
 
-        bool fReturningInternal = IsHDEnabled() && CanSupportFeature(FEATURE_HD_SPLIT) && fRequestedInternal;
+        bool fReturningInternal = fRequestedInternal;
+        fReturningInternal &= (IsHDEnabled() && CanSupportFeature(FEATURE_HD_SPLIT)) || IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS);
```
nit, add `HasHDSeed()` to complement `IsHDEnabled()`, and make it more clear what you're checking:
```cpp
// All WALLET_FLAG_DISABLE_PRIVATE_KEYS wallets were created after HD_SPLIT was added.
const bool fSupportsInternal = (HasHDSeed() && IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) || CanSupportFeature(FEATURE_HD_SPLIT)
bool fReturningInternal = fRequestedInternal & fSupportsInternal;
```

achow101 replied:
We can do that later.

Sjors commented on src/wallet/wallet.h:
```diff
@@ -993,6 +993,7 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     bool NewKeyPool();
     size_t KeypoolCountExternalKeys() EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
     bool TopUpKeyPool(unsigned int kpSize = 0);
+    void AddKeypoolPubkey(const CPubKey& pubkey, const bool internal);
```
Nit: call this new function from `TopUpKeyPool`, so it's more clear this function reuses existing code. It could take a `error_message` and an optional `WalletBatch` argument. If `WalletBatch` is set it should skip `NotifyCanGetAddressesChanged()`, though that gets a bit ugly.

achow101 replied:
I added a function `AddKeypoolPubkeyWithDB` which `TopUpKeypool` calls. `AddKeypoolPubkey` will make the `WalletBatch`, call `AddKeypoolPubkeyWithDB` and then ``NotifyCanGetAddressesChanged``.

Sjors reviewed (approved):
I'd still like to get this into 0.18 even if it's not tagged as such.

~tACK 8484a10~ (_see newer comment_) modulo `privateKeysDisabled` needs to be restored on the QT overview page (I can live with punting that).

For a followup I'd like to loosen the restriction of not adding imported private keys to the keypool. That's not necessary for [HWI](https://github.com/bitcoin-core/HWI), so can wait until after the branching. Implemented here: https://github.com/Sjors/bitcoin/commits/2019/02/add_privkey_to_keypool

Consider squashing dffff150e98fc96c98e9e8caf40cec7391bbebfb into 39c95977b2435110114cffd620a2efdb02357360

achow101 commented on test/functional/wallet_importmulti.py:
```diff
@@ -625,6 +625,74 @@ def run_test(self):
+            }]
+        )
+        assert result[0]['success']
+        assert result[1]['success']
```
Done

achow101 commented on src/qt/bitcoingui.cpp:
```diff
@@ -1202,7 +1202,7 @@ void BitcoinGUI::updateWalletStatus()
     WalletModel * const walletModel = walletView->getWalletModel();
     setEncryptionStatus(walletModel->getEncryptionStatus());
-    setHDStatus(walletModel->privateKeysDisabled(), walletModel->wallet().hdEnabled());
+    setHDStatus(!walletModel->canGetAddresses(), walletModel->wallet().hdEnabled());
```
I've reverted this change.

achow101 commented on src/qt/overviewpage.cpp:
```diff
@@ -161,7 +161,7 @@ void OverviewPage::setBalance(const interfaces::WalletBalances& balances)
     int unit = walletModel->getOptionsModel()->getDisplayUnit();
     m_balances = balances;
-    if (walletModel->privateKeysDisabled()) {
+    if (!walletModel->canGetAddresses()) {
```
I reverted that change

instagibbs commented on test/functional/wallet_importmulti.py:
```diff
@@ -625,6 +625,76 @@ def run_test(self):
                      ismine=False,
                      iswatchonly=False)
 
+        # Import some public keys to the keypool of a no privkey wallet
```
A test showing that importing some non-single-pubkey addresses using descriptors doesn't cause the keypool to grow seems apt.

achow101 replied:
Done

Sjors commented:
Yikes, I missed something in my earlier tACK, because I was too focussed on testing with private key support.

When I create a watch-only wallet, import keys and rescan, it produces the right addresses, but it's not picking up existing transactions, not detecting new incoming transactions.

```
importmulti '[{"desc": "sh(wpkh(tpubD6NzVbkrYhZ4YNXVQbNhMK1WqguFsUXceaVJKbmno2aZ3B6QfbMeraaYvnBSGpV3vxLyTTK9DYT1yoEck4XUScMzXoQ2U2oSmE2JyMedq3H/0h/0h/*))", "range": {"end": 2}, "timestamp": 1548979200, "keypool": true, "watchonly": true}]
```

```
getaddressinfo tb1q8vprchan07gzagd5e6v9wd7azyucksq2y38hpw
{
  "address": "tb1q8vprchan07gzagd5e6v9wd7azyucksq2y38hpw",
  "scriptPubKey": "00143b023c5fb37f902ea1b4ce985737dd11398b400a",
  "ismine": false,
  "solvable": true,
  "desc": "wpkh([3b023c5f]021ec1fdc1d288da27c93a79b0fb7f669314a6f477400d2a1019529f7ca2ee27b4)",
  "iswatchonly": false,
  "isscript": false,
  "iswitness": true,
  "witness_version": 0,
  "witness_program": "3b023c5fb37f902ea1b4ce985737dd11398b400a",
  "pubkey": "021ec1fdc1d288da27c93a79b0fb7f669314a6f477400d2a1019529f7ca2ee27b4",
  "label": "",
  "ischange": false,
  "labels": [
    {
      "name": "",
      "purpose": "receive"
    }
  ]
}
```

This looks correct to me (calling `deriveaddresses` on that inferred descriptor returns the address).

Remember when @instagibbs added:

> The second commit has watch keys imported to the keypool be marked as used when transactions are seen using them.

You then dropped that commit, because it was superseded by #14821.

The problem is that `AddToWalletIfInvolvingMe` only considers `IsMine` https://github.com/bitcoin/bitcoin/blame/master/src/wallet/wallet.cpp#L972. It's possible that some earlier version of #14021 incorrectly set IsMine so we didn't notice this while testing `hww`.

achow101 commented:
> But these imported public keys are merely `Solvable` so why aren't they skipped?

`IsMine()` returns true for things that are watch only, solvable, or spendable.

instagibbs commented:
More concretely, `IsMine` does a last-minute check if the exact scriptPubKey exists in the watchonly store if it fails at everything else.

instagibbs commented:
Feel free to remove "built on" comment in OP so I don't keep checking if it's time to review in earnest :)

instagibbs commented:
There is no test for keys being imported in order, and it appears to not actually be implemented for descriptors with ranges? You're iterating over a map's keys, which will be sorted in `CKeyID` order.

instagibbs commented:
Good news: I'm wrong, and I'm the one who wrote the logic originally :( 

Test and I'm happy.

instagibbs commented on test/functional/wallet_importmulti.py:
```diff
@@ -625,6 +625,119 @@ def run_test(self):
+        ]
+        result = wrpc.importmulti(
+            [{
+                'desc': 'wpkh(' + xpub + '/0h/0h/*' + ')',
```
```
./test/functional/wallet_importmulti.py:730:35: F821 undefined name 'xpub'
```

achow101 replied:
Whoops

Sjors commented:
@instagibbs keys from a ranged descriptor are imported in order. Feel free to replace this test:

```py
        # Test importing of a ranged descriptor without keys, check order
        self.log.info("Should import the ranged descriptor with specified range as solvable")
        self.test_importmulti({"desc": desc,
                               "timestamp": "now",
                               "range": {"end": 9}},
                              success=True,
                              warnings=["Some private keys are missing, outputs will be considered watchonly. If this is intentional, specify the watchonly flag."])
        # Check the first two addresses:
        for address in addresses:
            test_address(self.nodes[1],
                         key.p2sh_p2wpkh_addr,
                         solvable=True)

        # Check keys were imported in the correct order:
        for i in range(10):
            address = self.nodes[1].getnewaddress()
            assert_equal(self.nodes[1].getaddressinfo(address)["hdkeypath"], "m/0'/0'/%s'" % i)
```

instagibbs commented:
Test fix only change.

reACK https://github.com/bitcoin/bitcoin/pull/14075/commits/32fd3bb627fbd3c44b4f45434ead56f207dffa06

instagibbs commented:
re-ACK https://github.com/bitcoin/bitcoin/pull/14075/commits/38fbc101d1342222565b8670b7cd9c2485a57c74

Ordering test fixed, at least locally.

Sjors commented:
utACK a6e2c3331023d11112f8e9d176aa823d84b3f5ce (only adds a test since my last ACK)

MeshCollider commented:
utACK https://github.com/bitcoin/bitcoin/pull/14075/commits/38fbc101d1342222565b8670b7cd9c2485a57c74

MeshCollider commented:
LGTM https://github.com/bitcoin/bitcoin/pull/14075/commits/f4b00b70e811d3772589ab3c64d7658f4dbdab69

Sjors commented:
Post merge utACK f4b00b7
